core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -77,6 +77,11 @@ class BlockManagerMasterEndpoint(
   // Mapping from block id to the set of block managers that have the block.
   private val blockLocations = new JHashMap[BlockId, mutable.HashSet[BlockManagerId]]
 
+  // Mapping from task id to the set of rdd blocks which are generated from the task.
+  private val tidToRddBlockIds = new mutable.HashMap[Long, mutable.HashSet[RDDBlockId]]
+  // Record the visible RDD blocks which have been generated at least from one successful task.
+  private val visibleRDDBlocks = new mutable.HashSet[RDDBlockId]
```

Comment:
Instead of keeping track of visible blocks, should we be tracking those which are not yet visible ?
The size would be much smaller - blocks by default get added to this `Set`, and get promoted to visible explicitly.

Reply:
That would be better. Thanks.

Reply:
Just found one problem if we track the invisible RDD blocks.

If we track the invisible RDD blocks, then we would mark a RDD block as visible(cache can be used) only when it exists in `blockLocations` and not exists in `invisibleRDDBlocks`. When `blockLocations` removed the block(could be caused by executor lost), we will lose the information. Then the new cached data won't be leveraged as soon as possible(right after the cache is generated/reported to master).

Reply:
Not sure I follow - we cannot leverage the block until it is explicitly marked visible (unless within same task).
If the executor is lost before that happens, then we can never make it visible - and so fine to clean up from invisible blocks, right ?

Reply:
Let me explain this further. If we track visible blocks, it's clear that we always know which blocks are visible.

If we track invisible blocks, the way we consider a block as visible is that at least one block exists and it's not in invisible lists. So if the existing blocks got lost, we will lose the information. Next time the cache is re-computed, we will do this again(firstly put it into invisible lists, then promote it to visible by removing it from invisible list once task finished successfully). And after doing the process again, the cache would be visible then.

Reply:
Yes, that is the idea - block gets added to invisible list initially (since task has not completed), and then gets promoted to becoming visible (when task completes).

If existing blocks are lost - why would you need that information as they are gone ? In other words, how is it different from today's situation (without visibility) - if a block is lost, it is no longer in system.

Reply:
> If existing blocks are lost - why would you need that information as they are gone? In other words, how is it different from today's situation (without visibility) - if a block is lost, it is no longer in system.

Here is an example for the scenario I am trying to describe:
1. we have a cached block rdd_1_1 which has been successfully cached and marked as visible.
2. the cached block got lost due to executor lost;
3. another task on rdd1 got submitted and the 1st attempt failed after putting the cache block rdd_1_1, for the 2nd attempts, things could be different here:
    a. if we still have the visiblily status, the 2nd attempt can use the cached block directly;
    b otherwise, we still need to do the computing.

Reply:
A bit unclear on this actually ... why would a block status be cached on an executor (other than driver) which is not hosting the block ?
If executor is lost, driver will update its state as well due to the loss (evict the block) ... if there are other copies of the block, continues to be visible (since block exists). If task fails before block transitions to visible (or executor loss), driver will evict all blocks.

Note, when executor is fetching a block for computation from remote executor - it should be 'seeing' a block only if it is visible.
(If block is locally available due to replication - then it is invisible until marked visible. And so same logic as above would work).

Reply:
> If executor is lost, driver will update its state as well due to the loss (evict the block) ... if there are other copies of the block, continues to be visible (since block exists). If task fails before block transitions to visible (or executor loss), driver will evict all blocks.

+1

Reply:
Will try to make the change, the scenario I mentioned above would not be a common case.

Reply:
Updated. Please take a look. Thanks. cc @Ngone51  @mridulm